cedar-policy/src/ffi/format.rs

```diff
@@ -0,0 +1,164 @@
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//! JSON FFI entry points for the Cedar policy formatter. The Cedar Wasm
+//! formatter is generated from the [`format()`] function in this file.
+
+#![allow(clippy::module_name_repetitions)]
+
+use super::utils::DetailedError;
+use cedar_policy_formatter::{policies_str_to_pretty, Config};
+use serde::{Deserialize, Serialize};
+#[cfg(feature = "wasm")]
+use wasm_bindgen::prelude::wasm_bindgen;
+
+#[cfg(feature = "wasm")]
+extern crate tsify;
+
+/// Apply the Cedar policy formatter to a policy set in the Cedar policy format
+#[cfg_attr(feature = "wasm", wasm_bindgen(js_name = "formatPolicies"))]
+#[allow(clippy::needless_pass_by_value)]
+pub fn format(call: FormattingCall) -> FormattingAnswer {
+    let config = Config {
+        line_width: call.line_width,
+        indent_width: call.indent_width,
+    };
+    match policies_str_to_pretty(&call.policy_text, &config) {
+        Ok(prettified_policy) => FormattingAnswer::Success {
+            formatted_policy: prettified_policy,
+        },
+        Err(err) => FormattingAnswer::Failure {
+            errors: vec![err.into()],
+        },
+    }
+}
+
+/// Apply the Cedar policy formatter. Input is a JSON encoding of
+/// [`FormattingCall`] and output is a JSON encoding of [`FormattingAnswer`].
+///
+/// # Errors
+///
+/// Will return `Err` if the input JSON cannot be deserialized as a
+/// [`FormattingCall`].
+pub fn format_json(json: serde_json::Value) -> Result<serde_json::Value, serde_json::Error> {
+    let ans = format(serde_json::from_value(json)?);
+    serde_json::to_value(ans)
+}
+
+/// Apply the Cedar policy formatter. Input and output are strings containing
+/// serialized JSON, in the shapes expected by [`format_json()`].
+///
+/// # Errors
+///
+/// Will return `Err` if the input cannot be converted to valid JSON or
+/// deserialized as a [`FormattingCall`].
+pub fn format_json_str(json: &str) -> Result<String, serde_json::Error> {
+    let ans = format(serde_json::from_str(json)?);
+    serde_json::to_string(&ans)
+}
+
+/// Struct containing the input data for formatting
+#[derive(Serialize, Deserialize, Debug)]
+#[cfg_attr(feature = "wasm", derive(tsify::Tsify))]
+#[cfg_attr(feature = "wasm", tsify(into_wasm_abi, from_wasm_abi))]
+#[serde(rename_all = "camelCase")]
+#[serde(deny_unknown_fields)]
+pub struct FormattingCall {
+    /// Policy text. May define multiple policies or templates in the Cedar policy format.
+    policy_text: String,
+    /// Line width
+    line_width: usize,
+    /// Indentation width
+    indent_width: isize,
```

Comment:
should we be opinionated and set defaults for these?  Does the CLI interface to the formatter have defaults for these?

Reply:
The CLI defaults are 80 for line width & 2 for indentation. Will use the same defaults here.

Reply:
Done in the latest commit